model/stats/db.go

```diff
@@ -442,15 +443,309 @@ func statsToUpdatePipeline(projectId string, start time.Time, end time.Time) []b
+	if numDays <= 1 {
+		return bson.M{"$addFields": bson.M{fieldName: "$_id.date"}}
+	}
+	boundaries := dateBoundaries(start, end, numDays)
+	branches := make([]bson.M, len(boundaries))
+	for i := 0; i < len(boundaries)-1; i++ {
+		branches[i] = bson.M{
+			"case": bson.M{"$and": array{
+				bson.M{"$gte": array{"$_id.date", boundaries[i]}},
+				bson.M{"$lt": array{"$_id.date", boundaries[i+1]}},
+			}},
+			"then": boundaries[i],
+		}
+	}
+	lastIndex := len(boundaries) - 1
+	branches[lastIndex] = bson.M{
+		"case": bson.M{"$gte": array{"$_id.date", boundaries[lastIndex]}},
+		"then": boundaries[lastIndex],
+	}
+	return bson.M{"$addFields": bson.M{fieldName: bson.M{"$switch": bson.M{"branches": branches}}}}
+}
+
+// Builds the _id field for the $group stage corresponding to the GroupBy value.
+func buildGroupId(groupBy GroupBy) bson.M {
+	id := bson.M{"date": "$date"}
+	switch groupBy {
+	case GroupByDistro:
+		id["distro"] = "$_id.distro"
+		fallthrough
+	case GroupByVariant:
+		id["variant"] = "$_id.variant"
+		fallthrough
+	case GroupByTask:
+		id["task_name"] = "$_id.task_name"
+		fallthrough
+	case GroupByTest:
+		id["test_file"] = "$_id.test_file"
+	}
+	return id
+}
+
+// Edits a match expression to include that the 'fieldName' field must have its value in 'values'.
+// Does nothing if 'values' is empty.
+func addMatchIn(matchExpr bson.M, fieldName string, values []string) {
+	if len(values) == 1 {
+		matchExpr[fieldName] = values[0]
+	} else if len(values) > 1 {
+		matchExpr[fieldName] = bson.M{"$in": values}
+	}
+}
+
+// Adds to an existing $match expression the conditions imposed by the filter StartAt field
+func addMatchTestPagination(matchExpr bson.M, filter *StatsFilter) {
+	var dateOperator string
+	if filter.Sort == SortEarliestFirst {
+		dateOperator = "$gt"
+	} else {
+		dateOperator = "$lt"
+	}
+
+	var fields []string
+	var operators []string
+	var values []interface{}
+
+	switch filter.GroupBy {
+	case GroupByTest:
+		fields = []string{"_id.date", "_id.test_file"}
+		operators = []string{dateOperator, "$gt"}
+		values = []interface{}{filter.StartAt.Date, filter.StartAt.Test}
+	case GroupByTask:
+		fields = []string{"_id.date", "_id.task_name", "_id.test_file"}
+		operators = []string{dateOperator, "$gt", "$gt"}
+		values = []interface{}{filter.StartAt.Date, filter.StartAt.Task, filter.StartAt.Test}
+	case GroupByVariant:
+		fields = []string{"_id.date", "_id.variant", "_id.task_name", "_id.test_file"}
+		operators = []string{dateOperator, "$gt", "$gt", "$gt"}
+		values = []interface{}{filter.StartAt.Date, filter.StartAt.BuildVariant, filter.StartAt.Task, filter.StartAt.Test}
+	case GroupByDistro:
+		fields = []string{"_id.date", "_id.variant", "_id.task_name", "_id.test_file", "_id.distro"}
+		operators = []string{dateOperator, "$gt", "$gt", "$gt", "$gt"}
+		values = []interface{}{filter.StartAt.Date, filter.StartAt.BuildVariant, filter.StartAt.Task, filter.StartAt.Test, filter.StartAt.Distro}
+	}
```

Comment:
Is there anything that validates that we've specified the right number of arguments to each. It seems *really* easy to make a change here and get very different semantics. Seems like the operators are "whatever the date operator is plus $gt until the end.

Reply:
I created a struct to combine the field, operator and value, and now pass a array of the struct.